### PR TITLE
[MODULAR] Adds a Unit Test for Sprite Accessories

### PR DESCRIPTION
**Pull request by RimiNosha**
## About The Pull Request
I'm so done with our broken ass accessories system, so this will make the sprite system scream if literally anything goes slightly wrong now.

Fuck your logfiles, I don't care. I'm so done, this shit's the sole reason the synth rework has taken over ~~a week~~ two weeks.

I'll fix the stupid sprite names and any broken accessories once I've gotten comprehensible logs. For now, I'm going to sleep, I'm tired and in mental and physical pain.

No, I am not taking this upstream, this is for OUR system, which is fundamentally different in use from TGs one, and their one arguably isn't used enough to warrant a whole unit test.

Oh, also improves our sprite accessories to NOT generate up to 35 icons each. Yes, you heard that right. 35. Fucking. EACH.

You have the base 5 from the five layers.
Times that by up to 6, cause there's potentially matrixed colours which add three instead of one, plus extra1, extra2, and inner for each layer.
That's a total of 30. But wait, it gets better.
Add another five, because each layer gets an additional emissive overlay if enabled.
That's a grand total of 35. It's usually around 15, but there's a fair few cases where it's 30, ignoring the user-enabled emissives.

"Oh, that's for all of the sprite accessories added together, right?"
Hahahaha, oh my god you don't even know, that's PER sprite accessory. Times that by however many mutant parts (tails, ears, etc) and sprite accessories you added in the augments tab.
Oh boy, that's a big number, isn't it?
In the case of Rimi, who has just cat ears, a tail, and underwear, she has potentially 107 extra sprites being generated that shouldn't be there. These are the actual values I'm estimating from code. Underwear generated 87 of those.
Imagine what someone who actually uses the tattoos, markings, mutant parts and other similar systems generates!

"But doesn't it check to make sure the sprite exists?"
Fuck no it doesn't, cause that would be sensible, and we don't do that here.
There's not a single fucking check, it just yeets that value into the layers applier, and if it's an invalid icon, it's just left transparent. Good shit.

## How This Contributes To The Skyrat Roleplay Experience
Less angry messages in the contributor channel when I tear my hair out when lazy coders and spriters don't test their shit is good, yes?

Also should fix some sprite accessories not working at all in the first place, due to said lazy coders and spriters.

## Proof of Testing

This is a test that does testing for people before we have to deal with the consequences of them not doing it themselves. 

No, I'm not including the logs here, cause github actions has my back for that.

Here's some neat performance metrics, though, if a little squint cause I can't consistently generate identical numbers.
image
image
Old vs new, I was told that the 4th number mattered the most.

## Changelog

:cl:
add: Adds a new unit test which fails if any sprite accessory is unable to have an icon found.
code: Made the sprite accessories system scream in console whenever something isn't right instead of silently failing.
fix: Your sparkledogs with 30+ sprite accessories should no longer lag the server when you're editing or spawning as them.
refactor: All sprites have had a sanity check, and now only apply on layers they ACTUALLY should be applied, preventing cases where up to 35 images, per sprite accessory, are generated, when it should be only like 5.
fix: The broken Vox hairstyle no longer exists.
fix: The IPC Square Synth screen should now actually work.
/:cl:

### CODE REVIEW

**Comment by RimiNosha** [comment]
Leaving undrafted so tests run.

**Comment by Iamgoofball** [comment]
you should also include screenshot tests for this

**Comment by RimiNosha** [comment]
I have no intention to add the literal 100s of images that'd be required to do that properly.

**Comment by Iamgoofball** [comment]
you can generate them automatically when you're making the screenshot tests in the first place, but fair

**Comment by RimiNosha** [comment]
Drafting cause this one might take a bit to get just right.

**Comment by RimiNosha** [comment]
Oh my god, I have uncovered some downright disgusting shit that I'm gonna schizopost in the main post.

**Comment by RimiNosha** [comment]
Oh boy, I wonder how many errors this merge conflict will generate!

File: code/modules/unit_tests/~skyrat/sprite_accessories.dm [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// Yea, I'm doing this. This shit annoys me too much.
+// Set your layers properly, or be yelled at angrily. - Rimi
+/datum/unit_test/sprite_accessories
+	var/list/genitals_to_add = list(
```
Comment (Zonespace27) [inline code/modules/unit_tests/~skyrat/sprite_accessories.dm]: static

File: modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm [hunk 2]
```diff
@@ -98,98 +100,121 @@ GLOBAL_LIST_EMPTY(customizable_races)
...
+						else
+							accessory_overlay.color = "#AAA" //The gray husk color
 					else
-						accessory_overlay.color = "#AAA" //The gray husk color
+						switch(bodypart_accessory.color_src)
+							if(USE_ONE_COLOR)
+								accessory_overlay.color = mutant_bodyparts[key][MUTANT_INDEX_COLOR_LIST][1]
+							if(USE_MATRIXED_COLORS)
+								var/list/color_list = mutant_bodyparts[key][MUTANT_INDEX_COLOR_LIST]
+								for(var/n in color_layer_list)
```
Comment (Zonespace27) [inline modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm]: more descriptive var

File: modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm [hunk 3]
```diff
@@ -98,98 +100,121 @@ GLOBAL_LIST_EMPTY(customizable_races)
...
-					var/mutable_appearance/emissive_overlay = emissive_appearance_copy(accessory_overlay, owner)
-					//if (bodypart_accessory.center)
-					//	emissive_overlay = center_image(emissive_overlay, x_shift, bodypart_accessory.dimension_y)
-					standing += emissive_overlay
-
-			if(bodypart_accessory.hasinner)
+					accessory_overlay.color = override_color
+
+				if(accessories)
+					for (var/acces in accessories)
```
Comment (Zonespace27) [inline modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm]: could just do `standing += accessories`, pretty sure

File: modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -242,10 +270,11 @@ GLOBAL_LIST_EMPTY(customizable_races)
 							extra2_accessory_overlay.color = owner.hair_color
 
 				standing += extra2_accessory_overlay
-			if (specific_alpha != 255 && !override_color)
-				for (var/ov in standing)
+
+			if(specific_alpha != 255 && !override_color)
+				for(var/ov in standing)
```
Comment (Zonespace27) [inline modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm]:
```suggestion
				for(var/image/overlay as anything in standing)
```

File: modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm [hunk 5]
```diff
@@ -242,10 +270,11 @@ GLOBAL_LIST_EMPTY(customizable_races)
...
 					var/image/overlay = ov
```
Comment (Zonespace27) [inline modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm]:
```suggestion
```

File: modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -242,10 +270,11 @@ GLOBAL_LIST_EMPTY(customizable_races)
...
-					if (!istype(overlay.color,/list)) //check for a list because setting the alpha of the matrix colors breaks the color (the matrix alpha is set above inside the matrix)
+					if(!istype(overlay.color,/list)) //check for a list because setting the alpha of the matrix colors breaks the color (the matrix alpha is set above inside the matrix)
```
Comment (Zonespace27) [inline modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm]: `islist`

**Review by Zonespace27 — COMMENTED** [review]
happy to have more unit testing

File: modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm (hunk 3 above)
Comment (RimiNosha) [reply modular_skyrat/modules/customization/modules/mob/living/carbon/human/species.dm]: You are absolutely correct.

**Comment by RimiNosha** [comment]
I seem to have fuck some overlays. 

I'll fix it tomorrow, it's like 4am.

**Comment by RimiNosha** [comment]
I fucked up in a different way again. Sleep time before I get too mad.

**Comment by RimiNosha** [comment]
Yup, I'm done.

Giving up on this one for now. Someone else can pick it up if they want.

The current freeze, plus lack of patience, plus time between all my other projects, and possibly skill to continue trying to unfuck this system mount up to far outweigh my want for having this.

**Pull request closed without merging**